File: 04.builder/builder.java
class Order {
    private String product;
    private String address;
    private String deliveryDate;
    
    private Order(Builder builder) {
        this.product = builder.product;
        this.address = builder.address;
        this.deliveryDate = builder.deliveryDate;
    }

    public static class Builder {
        private String product;
        private String address;
        private String deliveryDate;

        public Builder(String product) {
            this.product = product;
        }
        public Builder setAddress(String address) {
            this.address = address;
            return this;
        }
        public Builder setDeliveryDate(String deliveryDate) {
            this.deliveryDate = deliveryDate;
            return this;
        }
        public Order build() {
            return new Order(this);
        }
    }
}
public class builder {
    public static void main(String[] args) {
        Order order = new Order.Builder("Laptop")
                            .setAddress("123 Main St")
                            .setDeliveryDate("10/10/2023")
                            .build();
    }
}

File: 04.builder/builder.py
class Order:
    def __init__(self, product):
        self.product = product
        self.address = None
        self.delivery_date = None

    def __str__(self):
        return f"Producto: {self.product}, Dirección: {self.address}, Fecha: {self.delivery_date}"

class OrderBuilder:
    def __init__(self, product):
        self.order = Order(product)

    def set_address(self, address):
        self.order.address = address
        return self

    def set_delivery_date(self, delivery_date):
        self.order.delivery_date = delivery_date
        return self

    def build(self):
        return self.order

# Uso
order = (OrderBuilder("Laptop")
         .set_address("123 Main St")
         .set_delivery_date("2023-10-10")
         .build())

print(order)
